### PR TITLE
Sort languages

### PR DESCRIPTION
**Pull request by mweimerskirch**
Currently, languages are listed in a more or less random order (presumably the order they were added in). With this patch, the languages are sorted alphabetically by locale in the source code for the purpose of making the code a bit cleaner.

In the UI, they are sorted alphabetically by localized name at runtime (second commit).

Example with English locale:
image

Example with German locale:
image

### CODE REVIEW

**Comment by yschroeder** [comment]
Haven't tested it yet, but it looks like a good idea.

Have you noticed that there are xlf files for other languages than we currently offer? These were generated by crowdin. I have no clue about their quality.

Should we add the other languages while we're at it?

**Comment by qcasey** [comment]
If this (and any future PR) breaks CI tests, should it include updates/fixes to those as well?

**Comment by shamoon** [comment]
> Should we add the other languages while we're at it?

Yea at the moment our workflow means that a new language will generate empty files until folks start translating. I think we should hold off on adding new languages in the UI until they’re at least mostly complete, see https://crowdin.com/project/paperless-ngx . In other words I’d leave out any that are 0-10% ish for now.

**Comment by shamoon** [comment]
The code itself looks OK but I will wait to approve until I actually test. 

@qcasey why does this break CI tests (I dont pretend to have a deep understanding of that stuff)?

**Comment by shamoon** [comment]
Oh I see, something to do with testing frontend languages https://github.com/paperless-ngx/paperless-ngx/blob/7eca68dfc06038ef1c8b7106ac595f892644e31e/src/documents/tests/test_views.py#L16 . Yea would be good to sort that.

**Comment by yschroeder** [comment]
In general I am a big fan of the idea that every single commit (or PR) has green tests. However, this can deter people from contributing.

But I also was not aware that there are tests in the backend realm that test for front end functionality. I guess I have to take a look what else we have there.

**Comment by shamoon** [comment]
This PR includes a change to a python file. Looking at it, if the frontend UI is sorting the languages do we even need to change the order in `settings.py` at all? As OP said that is just to make code “cleaner” but I’m not sure it really does anyway

**Comment by qcasey** [comment]
> In general I am a big fan of the idea that every single commit (or PR) has green tests. However, this can deter people from contributing.

Yeah that's a good point. Maybe it just needs to be fixed by someone on the team before merge, not necessarily by the contributor?

> This PR includes a change to a python file. Looking at it, if the frontend UI is sorting the languages do we even need to change the order in settings.py at all? As OP said that is just to make code “cleaner” but I’m not sure it really does anyway

+1 I imagine the changes to [src/paperless/settings.py](https://github.com/paperless-ngx/paperless-ngx/pull/78/files#diff-b3c4de627bca9188515497a784a0c368303d6c9a49a66d8dcc61d08f77ce3099) is all that's breaking CI anyway.

**Comment by shamoon** [comment]
Unless this PR is actually breaking something about the frontend serving the correct language (which is what that test is for). Sorry for the stream of consciousness, I will (or obviously others feel free) actually test this when I have some time tonight.

**Comment by qcasey** [comment]
Reverting the change to the python script fixes the tests, and the frontend still appears sorted as you predicted. I vote for reverting the `LANGUAGES = [` change. 

LGTM otherwise. 

Screenshot from 2022-02-17 09-26-56

**Comment by yschroeder** [comment]
Likely only the change to the ts file is needed. Angular.json should also stay untouched.

**Comment by shamoon** [comment]
> Likely only the change to the ts file is needed. Angular.json should also stay untouched.

Agree, if the goal here is just to sort languages for the user on the frontend then that’s all this PR should do.

**Review by shamoon — CHANGES_REQUESTED** [review]
@mweimerskirch as per discussion please revert changes from https://github.com/paperless-ngx/paperless-ngx/pull/78/commits/d3e1728b9e00ac8a4d361be284eb87461648522f

**Comment by mweimerskirch** [comment]
I took a break for a couple hours and was a bit overwhelmed by the number of messages.
Sorry for not noticing that the tests were failing, I haven't done that much with Python.

After a short investigation, I noticed that the first language in the list is used as a fallback language by the backend in case the cookie contains an unknown language. That's why I now slightly changed the PR by putting English first (US, then UK) and then the rest in alphabetical order. Now the tests pass, and it makes sense to keep English first, given that it's the project's base language. So the code is still clean.

The UI change (as seen in the screenshots) is unaffected by this.

> Should we add the other languages while we're at it?
I would do that in a separate PR. For some languages, we need to find out what the date format is, and IMHO there should be a rule how much (in percentage) should be translated before a language is enabled in the project. But that's a topic for another discussion.

> Agree, if the goal here is just to sort languages for the user on the frontend then that’s all this PR should do.
Actually, I had two goals: to provide a logical order in the backend (main language first and the rest ordered by locale to improve code quality) as well as in the frontend (alphabetically by localized name). I hope the PR can be accepted like this.

If this gets merged, I think the next step in improving translations would be to write a few lines in the CONTRIBUTING.md file, explaining how to add new languages. I can do that in another PR. I think this would help new translators.

**Comment by shamoon** [comment]
Yes, sorry for the many comments, I realize this is a relatively small PR, appreciate your effort.

Do others feel that re-sorting those arrays does 'improve code quality', especially considering `en-US` still has to be left first in `settings.py`? My opinion is to leave those alone and only include the change that sorts the languages, but Im fine of others think it should be this way. Again, not trying to be obsessive here!

**Comment by mweimerskirch** [comment]
@shamoon Remember that the list will (hopefully) grow in the future. If you look at the history of those files, you'll see that the order was an issue in the past. e.g. "pt-pt" was added after "it-it", but it was added above it in the code, not following any logic. 

Even minor, but incremental code quality improvements make everybody's life easier in the future.

**Comment by shamoon** [comment]
Alright, fair enough. Last question I promise--does `en-GB` need to be at the top too? I assume just `en-US`? If were going to be consistent perhaps lets move GB back to alphabetical and just put a comment next to en-US explaining why its first.

Thank you again for contributing

**Comment by mweimerskirch** [comment]
> Alright, fair enough. Last question I promise--does `en-GB` need to be at the top too? I assume just `en-US`? If were going to be consistent perhaps lets move GB back to alphabetical and just put a comment next to en-US explaining why its first.

Yes, I can change that. The comment is definitely a good idea, prevents others from making the same mistake I did. If this is alright for the others, I'll make that change tomorrow.

File: src/paperless/settings.py
Comment (shamoon) [inline src/paperless/settings.py:306]:
```suggestion
    ("en-us", _("English (US)")), # needs to be first to act as fallback language
    ("de-de", _("German")),
    ("en-gb", _("English (GB)")),
```

**Review by shamoon — CHANGES_REQUESTED** [review]
Thanks for your patience @mweimerskirch !

**Comment by mweimerskirch** [comment]
There you go, I made the change as discussed.

**Review by shamoon — APPROVED** [review]

**Pull request merged**
